syntax = "proto3";

package reddit.post;

import "comment.proto";
import "subreddit.proto";

//objects
message Image {
  optional string url = 1;
}

message Video {
  repeated string frames = 2;
}

message PostMeta {
  optional string title = 1;
  optional string text = 2;
  optional int32 score = 3;
  optional PostState state = 4;
  optional string published = 5;
  optional int64 ID = 6;
  optional ContentType type = 9;
  repeated comment.Comment comment = 10; 
  optional subreddit.Subreddit subreddit = 11;
  repeated string tags = 12;

  enum ContentType {
    IMAGE = 0;
    VIDEO = 1;
  }

  enum PostState {
    NORMAL = 0;
    LOCKED = 1;
    HIDDEN = 2;
  }
}


// requests/responses
message ImagePostResponse {
  PostMeta meta = 1;
  Image image = 2;
}

message VideoPostResponse {
  PostMeta meta = 1;
  Video video = 2;
}

message NewPostMeta {
  string title = 1;
  string text = 2;
  string state = 3;
  string subreddit = 4;
  repeated string tags = 5;
}

message NewImagePostRequest {
  NewPostMeta meta = 1;
  Image image = 2;
}

message NewVideoPostRequest {
  NewPostMeta meta = 1;
  Video video = 2;
}

message GetPostContentRequest {
  optional int64 postID = 1;
}

message GetPostContentResponse {
  optional PostMeta.ContentType type = 3;
  string videoframe = 1;
  string imageurl = 2;
}

message RatePostRequest {
  optional int64 postID = 1;
  optional string rating = 2;

  enum Rating {
    UPVOTE = 0;
    DOWNVOTE = 1;
  }
}

message RatePostResponse {
  optional PostMeta meta = 1;
}

message GetPostMetaRequest {
  optional int64 postID = 1;
}

message GetPostMetaResponse {
  optional PostMeta meta = 1;
}

message GetNCommentsRequest {
  optional int32 num_comments = 1;
  optional int64 postID = 2;
}

message GetNCommentsResponse {
  optional comment.Comment comment = 1;
  optional bool has_replies = 2;
}

message MonitorScoreRequest {
  optional int64 postID = 1;
  optional int64 commentID = 2;
}

message MonitorScoreResponse {
  optional int64 postID = 1;
  optional int32 postScore = 2;
  repeated int64 commentIDs = 3;
  repeated int32 commentScores = 4;
}

// services
service PostService {
  // send a title, a description, an initial state, and an image to be uploaded
  rpc PostImage (NewImagePostRequest) returns (ImagePostResponse) {}
  // send a title, a description, an initial state, and a video to be uploaded
  rpc PostVideo (NewVideoPostRequest) returns (VideoPostResponse) {}
  // send a post's id, get its content back
  rpc GetPostContent (GetPostContentRequest) returns (stream GetPostContentResponse) {}
  // send a post ID and a rating, get the new rating back with extraneous metadata
  rpc RatePost (RatePostRequest) returns (RatePostResponse) {}
  // send a post id, and get its metadata back
  rpc GetPost (GetPostMetaRequest) returns (GetPostMetaResponse) {}
  // send post ID and number of comments to get from it and get N most upvoted comments from it
  rpc GetNComments (GetNCommentsRequest) returns (stream GetNCommentsResponse) {}
  // send post ID to monitor post score. add comment ID to monitor comment scores in that post.
  // return stream of scores related to the IDs
  rpc MonitorPostScore (stream MonitorScoreRequest) returns (stream MonitorScoreResponse) {}
}
